File: Longest_Substring_Without_Repeating_Characters/solution2.java
import java.util.*;

class Solution {
    public int lengthOfLongestSubstring(String str) {
        final int L = str == null ? 0 : str.length();
        if (L == 0) return 0;
        
        int s = 0;
        int e = 0;
        int longest = 0;
        int[] map = new int[128];
        while (e < L) {
            char c = str.charAt(e);
            if (map[c] > s) s = map[c];
            longest = Math.max(longest, e - s + 1);
            map[c] = e + 1;
            e++;
        }
        return longest;
        
    }
}

File: Longest_Substring_Without_Repeating_Characters/solution2.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(str) {
    const L = str ? str.length : 0;
    if (L == 0) return 0;
    
    let s = 0;
    let e = 0;
    let longest = 0;
    let charMap = {};
    while (e < L) {
        let c = str[e];
        if (charMap[c] > s) s = charMap[c];
        longest = Math.max(longest, e - s + 1);
        charMap[c] = e + 1;
        e++;
    }
    return longest;
};
